### PR TITLE
Generate message code using md5()

### LINKED ISSUE #563: Changing non-alphanumeric characters does not change the string
Let's say, for example, I have the following string in one of my views:

```html
<p>{{ 'Choose one of the following:'|_ }}</p>
```

Then I later change my mind and decide to move the `:` outside the translation string:

```html
<p>{{ 'Choose one of the following'|_ }}:</p>
```

I then do a re-scan of the translate messages, however the string is still listed as `Choose one of the following:` with the `:` within the string, and when the translators see this, they will assume they need to include the `:` as well.

I see that in the `rainlab_translate_messages` table, the `code` column used looks like it removes all non-alphanumeric characters and replaces the spaces with a dot. Wouldn't it be more sensible to use a hash of the string instead of this? That way a single whitespace character change would trigger the whole string to be changed.

### PR DESCRIPTION
**Pull request by mjauvin**
- generate md5 hash of message id for message code
- includes migration file

fixes #563
replaces #564

Potential breaking change with Message Ids. Message Ids are now encoded using `md5()` to generate the Message code. Message Ids that contain characters other than letters, numbers, white spaces and `.` might have previously been sharing a Message translation due to their resulting Message code being equal. These sharing are no longer going to happen. Furthermore, if you were querying the Message model directly by code, those queries will now fail since the codes are now 32 hexadecimal characters (md5 hash of the Message Id).

### CODE REVIEW

**Comment by multiwebinc** [comment]
Wouldn't it be better to move `Message::makeMessageCode()` to the migration since that's the only place it should be needed now and should not need to be maintained within the actual Message class?

**Comment by multiwebinc** [comment]
`Str` is no longer needed in the `Message` model.

**Comment by multiwebinc** [comment]
Also, the test `testMakeMessageCode` should probably be removed.

**Comment by mjauvin** [comment]
> Also, the test `testMakeMessageCode` should probably be removed.

That was done in the last commit

File: tests/unit/models/MessageTest.php [hunk 1]
```diff
@@ -1,61 +0,0 @@
-<?php namespace RainLab\Translate\Tests\Unit\Models;
```
Comment (LukeTowers) [inline tests/unit/models/MessageTest.php]: The test should be rewritten not removed. Has @daftspunk provided input on this change?

Comment (mjauvin) [reply tests/unit/models/MessageTest.php]: No input on the change yet. I'll adapt the tests

File: models/Message.php [hunk 2]
```diff
@@ -276,23 +275,12 @@ protected static function makeCacheKey()
...
-        $messageId = preg_replace('!['.preg_quote('_').'|'.preg_quote('-').']+!u', $separator, $messageId);
-
-        // Remove all characters that are not the separator, letters, numbers, or whitespace.
-        $messageId = preg_replace('![^'.preg_quote($separator).'\pL\pN\s]+!u', '', mb_strtolower($messageId));
-
-        // Replace all separator characters and whitespace by a single separator
-        $messageId = preg_replace('!['.preg_quote($separator).'\s]+!u', $separator, $messageId);
-
-        return Str::limit(trim($messageId, $separator), 250);
+        return md5(trim($messageId));
```
Comment (mjauvin) [inline models/Message.php]: I wonder if the generated code should be case insensitive (as previous behavior)?
i.e. `return md5(trim(str_tolower($messageId)));`

Comment (multiwebinc) [reply models/Message.php]: What would the advantage of case-insensitivity be? And could there be a use case for case-sensitivity? For example, if the original author writes a string in all caps and then later decides that it looks silly that way, so they update the string. I would think that the translators should be aware of that.

Comment (multiwebinc) [reply models/Message.php]: Well, the same could be said for punctuation as well. People might have "My string!" in one place and "My string" in another place. I think that if you're going to create a change that could potentially break things, you might as well go all in so that the breaks occur on a single update. It's probably either now or never if you're going to change this.

Comment (mjauvin) [reply models/Message.php]: True, but I feel there is more potential for case sensitive ids than ones with punctuation.

In any case, we'll need @daftspunk's input on this breaking change.

Comment (multiwebinc) [reply models/Message.php]: @mjauvin Just a thought. What about adding a setting somewhere to set whether the messages are case-sensitive or not? Then the default could be set to insensitive and allow the user to set it as sensitive should the case arise.

**Comment by mjauvin** [comment]
@daftspunk do you see a problem merging this?

**Comment by mjauvin** [comment]
Getting ready to merge this for release 1.8.0, any objections?

I've tested it pretty well, and unit tests have been added as well.

**Comment by LukeTowers** [comment]
Isn't this a breaking change?

**Comment by mjauvin** [comment]
@LukeTowers It's not, unless multiple different messageIds resulted in the same code after characters other than  letters, numbers, white spaces and "." got removed in the old `makeMessageCode()` method.

For example, using the old `makeMessageCode` method, the following messageIds would have resulted in the same code:

- "my-message-key"
- "my=message=key"
- "my message key"
- "my.message.key"

This is no longer the case since we use md5() on the messageId.

**Comment by mjauvin** [comment]
@bennothommo any idea what's happening with the tests in here suddenly?

**Comment by LukeTowers** [comment]
The breaking change is in the changes made in the migration is it not? If someone is using `Message::whereCode()` manually, wouldn't that break as soon as the migration runs? @bennothommo what do you think, am I being too paranoid? Also does this affect the codes presented in the theme string translations section?

**Comment by bennothommo** [comment]
@LukeTowers it is a breaking change, but I think @mjauvin has appropriately updated it through making this an update that requires confirmation, so as long as people are made clearly aware of the change, we should be alright.

However - unless there's something I'm not considering - couldn't `whereCode` be adjusted to be compatible with previous usage scenarios? TBH, the Translate plugin is the plugin I use the least (ie. pretty much not at all) because I don't (and neither does my company) do multi-lang websites, so I'm not fully across it all.

**Comment by mjauvin** [comment]
@LukeTowers @bennothommo where do you see this `whereCode()` method?

@LukeTowers the theme string translations section uses messageIds (like in twig or in php code) and those get converted into a code by `makeMessageCode() `

**Comment by mjauvin** [comment]
@bennothommo @LukeTowers I don't understand why the tests are now failing, it makes no sense. What am I missing?

File: updates/version.yaml [hunk 3]
```diff
@@ -100,3 +100,7 @@
     - Polish localization updates
     - Turkish localization updates
     - Add Greek language localization
+1.8.0:
+    - "!!! Potential breaking change with Message Ids. Message Ids are now encoded using `md5()` to generate the Message code. Message Ids that contain characters other than letters, numbers, white spaces and `.` might have previously been sharing a Message translation due to their resulting Message code being equal. These sharing are no longer going to happen. Furthermore, if you were querying the Message model directly by code, those queries will now fail since the codes are now 32 hexadecimal characters (md5 hash of the Message Id)."
```
Comment (LukeTowers) [inline updates/version.yaml:104]:
```suggestion
    - "!!! Potential breaking change: Message codes are now MD5 hashed versions of the original string. See github.com/rainlab/translate-plugin/pull/556 for details."
```

File: models/Message.php [hunk 4]
```diff
@@ -289,23 +288,16 @@ protected static function makeCacheKey()
...
-
-        // Replace all separator characters and whitespace by a single separator
-        $messageId = preg_replace('!['.preg_quote($separator).'\s]+!u', $separator, $messageId);
-
-        return Str::limit(trim($messageId, $separator), 250);
+        if (Config::get('rainlab.translate::caseSensitiveIds', false)) {
+            return md5(trim($messageId));
+        } else {
+            return md5(mb_strtolower(trim($messageId)));
+        }
```
Comment (LukeTowers) [inline models/Message.php:301]: Why was this change required? Also wasn't the previous behavior to be case sensitive? Wouldn't you want these codes to be case sensitive?

Comment (mjauvin) [reply models/Message.php:301]: @LukeTowers previous behavior was case INSENSITIVE and is preserved by default. See old code below:
```
        $messageId = preg_replace('![^'.preg_quote($separator).'\pL\pN\s]+!u', '', mb_strtolower($messageId));
```

Comment (mjauvin) [reply models/Message.php:301]: The change is not REQUIRED but gives the option to developpers to change the behavior.

Comment (LukeTowers) [reply models/Message.php:301]: Why would we want to provide the option when we haven't up to this point? Personally I'd think that case sensitive should have been the default from the start, but now that it isn't, what's the benefit of allowing it to be?

Comment (mjauvin) [reply models/Message.php:301]: You kind of just answered that. If you think this should have been the default, why not give that option now? It doesn't hurt anything.

Comment (LukeTowers) [reply models/Message.php:301]: I'd be concerned about the message ID changing whenever the config item value changes

**Comment by mjauvin** [comment]
Good point. It shouldn't be changed once a project has started, but nothing can prevent it and it would cause mayhem... I'll revert this change.

**Comment by mjauvin** [comment]
The reason I was having trouble with the migration file during tests is that on the `down()` method, the table didn't exist anymore (not sure exactly why, but it's not the first time I see this behavior under the github test environment).

File: tests/unit/models/MessageTest.php [hunk 5]
```diff
@@ -11,51 +11,32 @@ public function testImportMessages()
...
-        $this->assertEquals('щука', Message::makeMessageCode('щука'));
-        $this->assertEquals('foo.漢字', Message::makeMessageCode('foo 漢字')); // Chinese
-        $this->assertEquals('xin.chào.thế.giới', Message::makeMessageCode('xin chào thế giới'));
-        $this->assertEquals('xin.chào.thế.giới', Message::makeMessageCode('XIN CHÀO THẾ GIỚI'));
-        $this->assertEquals('đấm.phát.chết.luôn', Message::makeMessageCode('đấm phát chết luôn'));
-        $this->assertEquals('foo', Message::makeMessageCode('foo ')); // no-break space (U+00A0)
-        $this->assertEquals('foo', Message::makeMessageCode('foo           ')); // spaces U+2000 to U+200A
-        $this->assertEquals('foo', Message::makeMessageCode('foo ')); // narrow no-break space (U+202F)
-        $this->assertEquals('foo', Message::makeMessageCode('foo ')); // medium mathematical space (U+205F)
-        $this->assertEquals('foo', Message::makeMessageCode('foo　')); // ideographic space (U+3000)
```
Comment (LukeTowers) [inline tests/unit/models/MessageTest.php:59]: Why remove all of these cases?

File: updates/migrate_message_code.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+class MigrateMessageCode extends Migration
+{
+    const TABLE_NAME = 'rainlab_translate_messages';
+
+    public function up()
+    {
+        if (!Schema::hasTable(self::TABLE_NAME)) {
+            return;
+        }
```
Comment (LukeTowers) [inline updates/migrate_message_code.php:16]: Why would this be required in the up() method?

File: updates/migrate_message_code.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+        // Remove all characters that are not the separator, letters, numbers, or whitespace.
+        $messageId = preg_replace('![^'.preg_quote($separator).'\pL\pN\s]+!u', '', mb_strtolower($messageId));
+
+        // Replace all separator characters and whitespace by a single separator
+        $messageId = preg_replace('!['.preg_quote($separator).'\s]+!u', $separator, $messageId);
+
+        return Str::limit(trim($messageId, $separator), 250);
+    }
+
```
Comment (LukeTowers) [inline updates/migrate_message_code.php:53]:
```suggestion
```

File: models/Message.php [hunk 8]
```diff
@@ -289,23 +288,12 @@ protected static function makeCacheKey()
     }
 
     /**
-     * Creates a sterile key for a message.
+     * Creates an md5 representation of the message Id
```
Comment (LukeTowers) [inline models/Message.php:291]:
```suggestion
     * Get the code for the provided message (32 character hexadecimal MD5-hashed version)
```

File: models/Message.php [hunk 9]
```diff
@@ -289,23 +288,12 @@ protected static function makeCacheKey()
...
      * @param  string $messageId
```
Comment (LukeTowers) [inline models/Message.php:292]:
```suggestion
     * @param  string $message
```

File: models/Message.php [hunk 10]
```diff
@@ -289,23 +288,12 @@ protected static function makeCacheKey()
...
      * @return string
      */
-    protected static function makeMessageCode($messageId)
+    public static function makeMessageCode($messageId)
```
Comment (LukeTowers) [inline models/Message.php:295]:
```suggestion
    public static function makeMessageCode($message)
```

File: models/Message.php [hunk 11]
```diff
@@ -289,23 +288,12 @@ protected static function makeCacheKey()
...
-        $messageId = preg_replace('!['.preg_quote('_').'|'.preg_quote('-').']+!u', $separator, $messageId);
-
-        // Remove all characters that are not the separator, letters, numbers, or whitespace.
-        $messageId = preg_replace('![^'.preg_quote($separator).'\pL\pN\s]+!u', '', mb_strtolower($messageId));
-
-        // Replace all separator characters and whitespace by a single separator
-        $messageId = preg_replace('!['.preg_quote($separator).'\s]+!u', $separator, $messageId);
-
-        return Str::limit(trim($messageId, $separator), 250);
+        return md5(mb_strtolower(trim($messageId)));
```
Comment (LukeTowers) [inline models/Message.php:297]:
```suggestion
        return md5(mb_strtolower(trim($message)));
```

File: tests/unit/models/MessageTest.php (hunk 5 above)
Comment (mjauvin) [reply tests/unit/models/MessageTest.php:59]: Because they are no longer relevant.

Comment (mjauvin) [reply tests/unit/models/MessageTest.php:59]: The tests were there to make sure the previous encoding was doing what it was supposed to.

Comment (LukeTowers) [reply tests/unit/models/MessageTest.php:59]: Would they still pass so long as the MD5 versions were used for comparison?

Comment (mjauvin) [reply tests/unit/models/MessageTest.php:59]: They would, but what's the point of comparing the md5 hash of a string with the result of a method which does pretty much only return the md5 hash of its argument?
